code/network/multimsgs.cpp

```diff
@@ -7676,74 +7768,111 @@ void process_NEW_countermeasure_fired_packet(ubyte *data, header *hinfo)
+	Assert(fire_info->shooter != nullptr);
+	Assert(fire_info->turret != nullptr);
+	Assert(override != nullptr);
+
+	if ( (fire_info->shooter == nullptr) || (fire_info->turret == nullptr) || (override == nullptr) ) {
 		return;
 	}
 
-	if (!(bfi_flags & BFIF_IS_FIGHTER_BEAM)) {
-		Assert(target != NULL);
-		if (target == NULL) {
+	if (fire_info->bfi_flags & BFIF_IS_FIGHTER_BEAM) {
+		Assertion(fire_info->bank > 0, "Bank for fighter BEAM is invalid!");
+		Assertion(fire_info->point > 0, "Bank point for fighter BEAM is invalid!");
+
+		if ( (fire_info->bank < 0) || (fire_info->point < 0) ) {
 			return;
 		}
-	}
+	} else {
+		Assertion(fire_info->target != nullptr, "BEAM fired without a target!");
 
-	target_sig = (target) ? target->net_signature : 0;
+		if (fire_info->target == nullptr) {
+			return;
+		}
 
-	u_beam_info = (short)beam_info_index;
+		shooter_subsys_index = ship_get_index_from_subsys(fire_info->turret, OBJ_INDEX(fire_info->shooter));
 
-	if (bfi_flags & BFIF_IS_FIGHTER_BEAM) {
-		Assert( (bank_point >= 0) && (bank_point < UCHAR_MAX) );
-		subsys_index = (char)bank_point;
-	} else {
-		subsys_index = (char)ship_get_index_from_subsys(turret, OBJ_INDEX(shooter));
-	}
+		Assertion(shooter_subsys_index >= 0, "BEAM fired from unknown subsystem!");
+		Assertion(shooter_subsys_index < SHRT_MAX, "BEAM fired from a subsystem beyond max limits!");
 
-	Assert(subsys_index >= 0);
-	if (subsys_index < 0) {
-		return;
+		if (shooter_subsys_index < 0) {
+			return;
+		}
 	}
 
-	// swap the beam_info override info into little endian byte order
-	b_info.dir_a.xyz.x = INTEL_FLOAT(&override->dir_a.xyz.x);
-	b_info.dir_a.xyz.y = INTEL_FLOAT(&override->dir_a.xyz.y);
-	b_info.dir_a.xyz.z = INTEL_FLOAT(&override->dir_a.xyz.z);
-
-	b_info.dir_b.xyz.x = INTEL_FLOAT(&override->dir_b.xyz.x);
-	b_info.dir_b.xyz.y = INTEL_FLOAT(&override->dir_b.xyz.y);
-	b_info.dir_b.xyz.z = INTEL_FLOAT(&override->dir_b.xyz.z);
-
-	b_info.delta_ang = INTEL_FLOAT(&override->delta_ang);
-	b_info.shot_count = override->shot_count;
+	target_sig = (fire_info->target) ? fire_info->target->net_signature : 0;
 
-	for (int i = 0; i < b_info.shot_count; i++) {
-		b_info.shot_aim[i] = INTEL_FLOAT(&override->shot_aim[i]);
+	if (fire_info->target && fire_info->target_subsys) {
+		target_subsys_index = ship_get_index_from_subsys(fire_info->target_subsys, OBJ_INDEX(fire_info->target));
 	}
 
+	u_beam_info = static_cast<short>(fire_info->beam_info_index);
+
 	// build the header
 	BUILD_HEADER(BEAM_FIRED);
-	ADD_USHORT(shooter->net_signature);
-	ADD_DATA(subsys_index);
+
+	ADD_USHORT(fire_info->shooter->net_signature);
 	ADD_USHORT(target_sig);
 	ADD_SHORT(u_beam_info);
-	ADD_DATA(b_info);  // FIXME: This is still wrong, we shouldn't be sending an entire struct over the wire - taylor
-//	ADD_DATA(bfi_flags);	// this breaks the protocol but is here in case we decided to do that in the future - taylor
-//	ADD_DATA(target_pos);	// ditto - Goober5000
+	ADD_FLOAT(fire_info->accuracy);
```

Comment:
Accuracy is only used to generate `shot_aim` so it shouldn't be necessary to send.

Reply:
Gotcha, I'll remove that.